designs/spot-consolidation.md

```diff
@@ -0,0 +1,105 @@
+# Spot Consolidation
+
+## Problem Statement
+
+Customers face increased costs due to underutilized spot instances. A customer with 4 x 24cpu pods would launch a 24xl (96cpu) instance type. If three of the four pods complete, the instance will indefinitely run at 25% utilization until the pod completes, is able to fit elsewhere in the cluster, or additional work is scheduled to the node. Customers have raised this inefficiency as a critical pain point with Karpenter: https://github.com/aws/karpenter-core/issues/763. 
+
+Karpenter’s consolidation algorithm minimizes the price of EC2 instances given pod scheduling constraints. However, price minimization presents challenges for EC2 Spot, which offers a tradeoff between price and interruption rate. Karpenter uses EC2 Spot’s PriceCapacityOptimized (PCO) allocation strategy to provide a balanced approach to this tradeoff, and doesn't always select the cheapest instance type, since it may result in immediate interruption. If Karpenter were to consolidate spot capacity, it would repeatedly interrupt the same instance, walking down the PCO decision ladder, until only the lowest price remained. Because of these complications, Karpenter does not currently support "spot replacement" consolidation.
+
+All consolidation actions contain an inherent tradeoff between price and availability. Consolidation actions are equivalent to a spot interruption, but without the deadline. This tradeoff applies to on-demand capacity, as well. This document explores approaches to enable consolidation for spot instances without regressing to lowest price. 
+
+## Design Options
+
+### 1: Minimum Flexibility [Recommended] 
```

Comment:
Not sure that this needs to be touched on in this design doc (just due to scoping) but we should think about how to handle cases where the user hasn't prescribed enough flexibility on their NodePool to begin with, which would effectively completely disable consolidation

Reply:
🤔 , I wonder if we could follow this line of reasoning to make the consolidation logic agnostic to this spot problem, and contain it within the AWS provider. e.g.

1. AWS Cloud Provider could reject a spot request w/ <20 flexibility (configurable?)
2. During a consolidation pre-spin, this would appear as a `InsufficientCapacityError`
3. Consolidation could remove the candidate from the pool. 
4. OD launches would be unaffected.

Reply:
Thinking on this, I think this is a bad idea -- you really need a way to communicate the minimum flexibility to the binpacker, since it's ultimately the actor constraining the space of instance types.

I wonder if we should consider modeling this as `offering.requiredFlexibility`. e.g., an offering is only valid if the request is sufficiently flexible. The cloud provider could choose how to model flexibility for any of its offerings (e.g. spot/od, or specific instance types).

Reply:
FWIW, this could even be seen as an implementation detail of option 1, to answer the following question:

"How is minimum flexibility extensible to different cloud providers"

Reply:
I'm not sure that you want to model this as a minimum flexibility. I think you probably want to model it as a maximum flexibility in some cases and a minimum flexibility in others. For launches, you want it to be a maximum flexibility since you don't want to walk down the ladder after you launch, so you want to constrain the options that you launch with to some maximal flexibility (like 20). 

The other thing to keep in mind here is I think that you still want to _launch_ instances that are under the flexibility threshold, you just know that you won't be able to consolidate them after they are under the threshold. 

Obviously, for consolidation, we are talking about a minimum flexibility. That is -- the minimum required instance types that need to come back from the scheduling call in order for the consolidation operation to be able to proceed. 

It almost seems like we need a general "flexibility" concept on instance types at a minimum that can capture both of these angles. We might even be able to get away with hard-coding it on the outset but we need to hear back from other cloudproviders if there is a good starting point that we can work from that we could then add as a parameter later.

Reply:
There's also a bit of a question here on how this affects things and your ability to consolidate down the line if you pack things into a larger spot instance, which means that you would necessarily be under the flexibility metric on launch. My assumption is that this is still alright since you just won't consolidate until the instance gets small enough that the instance flexibility gets large enough to consolidate. The only down-side of this approach is that you may be interrupted at a higher rate depending on the availability of the instance types that you launch with under the fleixibility.

Reply:
This might be a better route to go, though, then outright rejecting users that are working under this flexibility threshold today for launches.